client/relay-chain-rpc-interface/src/rpc_client.rs

```diff
@@ -0,0 +1,454 @@
+		Error as JsonRpseeError,
+	},
+	rpc_params,
+	types::ParamsSer,
+	ws_client::WsClientBuilder,
+};
+use parity_scale_codec::{Decode, Encode};
+use sc_client_api::StorageData;
+use sc_rpc_api::{state::ReadProof, system::Health};
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+use sp_core::sp_std::collections::btree_map::BTreeMap;
+use sp_runtime::DeserializeOwned;
+use sp_storage::StorageKey;
+use std::sync::Arc;
+
+pub use url::Url;
+
+const LOG_TARGET: &str = "relay-chain-rpc-client";
+
+const NOTIFICATION_CHANNEL_SIZE_LIMIT: usize = 20;
+
+/// Client that maps RPC methods and deserializes results
+#[derive(Clone)]
+pub struct RelayChainRPCClient {
+	/// Websocket client to make calls
+	ws_client: Arc<JsonRPCClient>,
+
+	/// Retry strategy that should be used for requests and subscriptions
+	retry_strategy: ExponentialBackoff,
+
+	to_worker_channel: Option<TracingUnboundedSender<NotificationRegisterMessage>>,
+}
+
+/// Worker messages to register new notification listeners
+#[derive(Clone, Debug)]
+pub enum NotificationRegisterMessage {
+	RegisterBestHeadListener(Sender<PHeader>),
+	RegisterImportListener(Sender<PHeader>),
+	RegisterFinalizationListener(Sender<PHeader>),
+}
+
+/// Worker that should be used in combination with [`RelayChainRPCClient`]. Must be polled to distribute header notifications to listeners.
+pub struct RPCStreamWorker {
+	// Communication channel with the RPC client
+	client_receiver: TracingUnboundedReceiver<NotificationRegisterMessage>,
+
+	// Senders to distribute incoming header notifications to
+	imported_header_listeners: Vec<Sender<PHeader>>,
+	finalized_header_listeners: Vec<Sender<PHeader>>,
+	best_header_listeners: Vec<Sender<PHeader>>,
+
+	// Incoming notification subscriptions
+	rpc_imported_header_subscription: Subscription<PHeader>,
+	rpc_finalized_header_subscription: Subscription<PHeader>,
+	rpc_best_header_subscription: Subscription<PHeader>,
+}
+
+/// Entry point to create [`RelayChainRPCClient`] and [`RPCStreamWorker`];
+pub async fn create_worker_client(
+	url: Url,
+) -> RelayChainResult<(RPCStreamWorker, RelayChainRPCClient)> {
+	let mut client = RelayChainRPCClient::new(url).await?;
+	let best_head_stream = client.subscribe_new_best_heads().await?;
+	let finalized_head_stream = client.subscribe_finalized_heads().await?;
+	let imported_head_stream = client.subscribe_imported_heads().await?;
+
+	let (worker, sender) =
+		RPCStreamWorker::new(imported_head_stream, best_head_stream, finalized_head_stream);
+	client.set_worker_channel(sender);
+	Ok((worker, client))
+}
+
+fn handle_event_distribution(
+	event: Option<Result<PHeader, JsonRpseeError>>,
+	senders: &mut Vec<Sender<PHeader>>,
+) -> Result<(), String> {
+	match event {
+		// If sending fails, we remove the sender from the list because the receiver was dropped
+		Some(Ok(header)) => {
+			senders.retain_mut(|e| e.try_send(header.clone()).is_ok());
+			Ok(())
+		},
```

Comment:
The `try_send` can also return an error when the sender's channel is full, not only when it was dropped. (Since it's a bounded channel.)

Reply:
I added a error print and do not remove the sender for now. Alternative would be to make the task fail.

Reply:
Yeah, I think we need to think about what we should do if that happens and handle it appropriately, but considering this is still experimental and a work in progress I'm fine with it being as-is. But we'll need to revisit this issue before actually pushing this into production. Could you at least put like a TODO here as a reminder?